### PR TITLE
Studio: Snapshot counter not updating in settings when a site is deleted

### PR DESCRIPTION
**Pull request by kozer**
Fixes https://github.com/Automattic/dotcom-forge/issues/8132

## Proposed Changes

This diff fixes a bug when the snapshot counter in user settings is not updated when the user deletes a site from settings tab.

NOTES:
1. The problem was created due to circular dependencies between the different hooks and `useSiteDetails`.
2. The issue is now fixed, by centralizing all snapshot logic in one place, and use it in useSiteDetails, where necessary
3. It can simplified further, and moved into different hooks to offload the main `use-snapshots` provider. However, as we have moved to another project, I leave it for a different ticket.

## Testing Instructions
1. Create a site
5. Add a snapshot
6. Open user settings and ensure the counter has been increased
7. Navigate to settings
8. Press "Delete site" button at the bottom of the page
9. Ensure that after visiting settings, the counter has been decreased.

## Pre-merge Checklist

- [ ] Have you checked for TypeScript, React or other console errors?

### CODE REVIEW

File: src/ipc-handlers.ts [hunk 1]
```diff
@@ -524,7 +524,7 @@ export async function openLocalPath( _event: IpcMainInvokeEvent, path: string )
 export async function getThemeDetails( event: IpcMainInvokeEvent, id: string ) {
 	const server = SiteServer.get( id );
 	if ( ! server ) {
-		throw new Error( 'Site not found.' );
+		return null;
```
Comment (sejas) [inline src/ipc-handlers.ts]: I think this change is not related to this PR.

File: src/hooks/use-snapshots.tsx [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+		}
+	}, [ allSnapshots, clearFloatingSnapshots, initiated, loadingServerSnapshots ] );
+
+	// Save snapshots to storage when they change
+	useEffect( () => {
+		if ( ! initiated ) {
+			return;
+		}
+		getIpcApi().saveSnapshotsToStorage( snapshots );
+	}, [ snapshots, initiated ] );
```
Comment (sejas) [inline src/hooks/use-snapshots.tsx]: This code is duplicated in line 99:

https://github.com/Automattic/studio/pull/344/files#diff-494179dc77bcc1f7a2f661db06953be2b984040d47401a3eae9e9f6254385cbcR99-R104

```suggestion
```

File: src/hooks/use-snapshots.tsx [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+
+	useLayoutEffect( () => {
+		getIpcApi()
+			.getSnapshots()
+			.then( ( storedSnapshots ) => {
+				setSnapshots( storedSnapshots );
+				setInitiated( true );
+			} );
+	}, [] );
```
Comment (sejas) [inline src/hooks/use-snapshots.tsx:118]: Out of curiosity,  What's the benefit of using `useLayoutEffect` ?

File: src/hooks/use-snapshots.tsx [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+	isLoadingSnapshotUsage: boolean;
+	initiated: boolean;
+}
+
+export interface SnapshotStatusResponse {
+	is_deleted: string;
+	domain_name: string;
+	atomic_site_id: string;
+	status: '1' | '2';
+}
```
Comment (sejas) [inline src/hooks/use-snapshots.tsx]: What do you think about using an enum for the status?

```suggestion
enum SnapshotStatus {
	Deleted = '1',
	Active = '2',
}

export interface SnapshotStatusResponse {
	is_deleted: string;
	domain_name: string;
	atomic_site_id: string;
	status: SnapshotStatus;
}
```

File: src/hooks/use-archive-site.ts [hunk 5]
```diff
@@ -7,18 +7,25 @@ import { getIpcApi } from '../lib/get-ipc-api';
...
-import { SnapshotStatusResponse } from './use-delete-snapshot';
 import { useSiteDetails } from './use-site-details';
+import { useSnapshots } from './use-snapshots';
+
+interface SnapshotStatusResponse {
+	is_deleted: string;
+	domain_name: string;
+	atomic_site_id: string;
+	status: '1' | '2';
+}
```
Comment (sejas) [inline src/hooks/use-archive-site.ts]: This type is already declared in `src/hooks/use-snapshots.tsx`. Let's import it.

```suggestion
import { SnapshotStatusResponse, useSnapshots } from './use-snapshots';
```

**Review by sejas — APPROVED** [review]
I've tested it by creating, updating and deleting some snapshots, and checking the quota. Everything worked as expected.
I've added some optimizations and suggestions removing duplicated code.
Screenshot 2024-07-11 at 10 41 08

Screenshot 2024-07-11 at 10 41 04

File: src/hooks/use-snapshots.tsx (hunk 3 above)
Comment (kozer) [reply src/hooks/use-snapshots.tsx:118]: `useEffect` runs on initial render, and `useLayoutEffect` runs before that. As this is just to retrieve data, and better have them available when the UI is getting rendered, I found it to be better.
TL;DR: Get our data faster, before the initial render.

File: src/ipc-handlers.ts (hunk 1 above)
Comment (kozer) [reply src/ipc-handlers.ts]: I got an error at some point, before the refactor, when I tried different things to make the previous code work, so I decided to leave it. I'll move it back, as it was before and re test

Comment (kozer) [reply src/ipc-handlers.ts]: It seems, it's not needed anymore, so I removed it. I guess it has crashing the app due to race conditions.
